### ISSUE TITLE
pre-processing images

### ISSUE THREAD

**Issue by alexge233** [leonardvandriel/caffe2_cpp_tutorial#7]
Hi,

I'm trying to understand how things have changes from caffe to caffe2, and I've opened an [issue there](https://github.com/caffe2/caffe2/issues/987) but I have a few separate questions for this repository,
and in specific the [pretrained example](https://github.com/leonardvandriel/caffe2_cpp_tutorial/blob/master/src/caffe2/binaries/pretrained.cc), especially lines 55 to 87.

- I see you calculate and normalise each image, regardless of the network (which used to be the case?)
- You split the data in three separate channels (B, G, R?) but then you append it all in one vector?
- are those tensors single dimensional because of the network, or am I misunderstanding it?
- how can I upload the data to a GPU tensor since I'm doing all the opencv pre-processing using cuda?

Many thanks (and have a nice weekend!),
Alex

**Comment by leonardvandriel** [comment]
Perhaps it's helpful to summarize the steps taken in `pretrained.cc`:

1. Scale the image so that the shortest side of the image equals `FLAGS_size_to_fit`.
2. Crop to the image center so that both width and height equal `FLAGS_size_to_fit`.
3. Convert image color values from uchar (`CV_8UC3`) to float (`CV_32FC3`), assuming 3 color channels, and subtracting 128, so the mean color value is closer to 0.
4. Converting from NHWC to NCHW, which you could interpret as going from 1 image with 3 BGR colors per pixel to 3 BGR images with one color per pixel.

To answer your questions:

> I see you calculate and normalise each image, regardless of the network (which used to be the case?)

That's correct. I'm making a lot of assumptions here, which might cause problems. However most ImageNet models work best on square images of size 224x224, with mean float values between -128 and 128, 3 color channels in NCHW order.

> You split the data in three separate channels (B, G, R?) but then you append it all in one vector?
> are those tensors single dimensional because of the network, or am I misunderstanding it?

So that's the NCHW order, which is default with Caffe2. Getting the order wrong usually turns the image into noise resulting in predictions like "window screen". Take a look at [Caffe Prefers CHW Order](https://caffe2.ai/docs/tutorial-image-pre-processing.html#null__caffe-prefers-chw-order) for a more detailed explanation.

> how can I upload the data to a GPU tensor since I'm doing all the opencv pre-processing using cuda?

In my experience it's not necessary to store input data in a `TensorCUDA`, assuming this is copied automatically by the operator. However you can copy a `TensorCPU` into a `TensorCUDA` blob like this:

    TensorCPU tensor_cpu(...);
    TensorCUDA tensor_cuda = workspace.CreateBlob("input")->GetMutable<TensorCUDA>();
    tensor_cuda->ResizeLike(tensor_cpu);
    tensor_cuda->ShareData(tensor_cpu);

**Comment by alexge233** [comment]
@leonardvandriel Many thanks, I haven't had time yet to go through all that, but once I will I'd like to push a pre-processing example if that is OK?

The reason why I am asking about TensorCUDA is because I do all the pre-processing on the GPU, so I don't see the reason to upload the image on the device then copy it on the host only to copy it back to the device again.

I'll get back to you on this tomorrow.

**Comment by leonardvandriel** [comment]
Can you explain in more detail what you mean by "pre-processing on the GPU"? There's indeed no point in copying things to and from the host, but it don't think that's the case in any of the examples.

**Comment by alexge233** [comment]
No your examples don't use the GPU, the [issue](https://github.com/caffe2/caffe2/issues/987) I've posted in caffe2 uses OpenCV's `cuda` versions and the `GpuMat` matrix, which is being used more commonly with opencv >= 3.0

**Comment by leonardvandriel** [comment]
got it, thanks for explaining that. I've been using the older OpenCV 2.4, which was default on homebrew. Would be nice to have an example that uses `GpuMat`.

**Comment by alexge233** [comment]
I'll try to do a pull request as soon as I've finished this.
I've asked on [caffe2](https://github.com/caffe2/caffe2/issues/1004) how to set a `Tensor<CUDAContext>` because I can't find any examples or instructions, in either the source code or else where

**Comment by alexge233** [comment]
Hi Leonard, I'm back with some code.
Sadly it seems to throw an assertion of wrong dimensions. I'm copy-pasting, assuming you may be able to help. Feel free to add to the repository, or I can do a pull request if you prefer.

```cpp
caffe2::TensorCPU make_caffe_tensor::operator()(
                                                 const cv::cuda::GpuMat & image,
                                                 unsigned int channels,
                                                 cv::Size geometry
                                               )
{
    if (image.empty()) {
        throw std::runtime_error("empty image param");
    }
    // resize to input geometry
    cv::cuda::GpuMat resized;
    if (image.size() != geometry) {
        cv::cuda::resize(image, resized, geometry);
    }
    // crop image if needed
    cv::Rect crop((resized.cols - geometry.height) / 2,
                  (resized.rows - geometry.width) / 2, 
                  geometry.height,
                  geometry.width);

    // convert colour type if needed
    cv::cuda::GpuMat sample;
    if (resized.channels() == 3 && channels == 1) {
        cv::cuda::cvtColor(resized, sample, cv::COLOR_BGR2GRAY);
    }
    else if (resized.channels() == 4 && channels == 1) {
        cv::cuda::cvtColor(resized, sample, cv::COLOR_BGRA2GRAY);
    }
    else if (resized.channels() == 4 && channels == 3) {
        cv::cuda::cvtColor(resized, sample, cv::COLOR_BGRA2BGR);
    }
    else if (resized.channels() == 1 && channels == 3) {
        cv::cuda::cvtColor(resized, sample, cv::COLOR_GRAY2BGR);
    }
    else {
        sample = resized;
    }
    // convert to float and normalize in one operation
    cv::cuda::GpuMat mfloat;
    if (channels == 3) {
        sample.convertTo(mfloat, CV_32FC3, 1.0, -128);
    }
    else {
        sample.convertTo(mfloat, CV_32FC1, 1.0, -128);
    }
    // split to channels
    std::vector<cv::cuda::GpuMat> input_channels;
    cv::cuda::split(mfloat, input_channels);
    // allocate => channels * float matrix size
    std::vector<float> data(input_channels.size() * 
                            (mfloat.size().height * mfloat.size().width));
    // iterate, download and append 
    for (auto & mat : input_channels) {
        cv::Mat tmp(mat.size(), mat.type());
        mat.download(tmp); 
        data.insert(data.end(), (float *)tmp.datastart, (float *)tmp.dataend);
    }
    std::vector<caffe2::TIndex> dims({1, sample.channels(), sample.rows, sample.cols});
    caffe2::TensorCPU tensor(dims, data, NULL);
    return tensor;
}
```

My problem is that the above throws an assertion from the tensor constructor:

```
[enforce fail at tensor.h:128] values.size() == size_. 301056 vs 150528
```

AFAIK this is an issue of miss-matching dimensions?

**Comment by alexge233** [comment]
Hi @leonardvandriel 

I've managed to copy GPU pre-processed data from `cv::cuda` functions to a `TensorCUDA` but from what I can see, the `Predictor` only accepts `TensorCPU`. 
Do you know how I can force the `Predictor` to run a CUDA tensor instead?

**Comment by leonardvandriel** [comment]
The `Predictor` is only a convenience class for setting up a workspace and running the model, see [predictor.cc](https://github.com/caffe2/caffe2/blob/master/caffe2/core/predictor.cc). I believe the use of `Predictor` is even discouraged.

Take a look at `predict_example` in [mnist.cc](https://github.com/leonardvandriel/caffe2_cpp_tutorial/blob/d5daf4e89b9b1a12979649e783d68f4c821a554f/src/caffe2/binaries/mnist.cc#L380), which implements a predictor from scratch.

**Comment by alexge233** [comment]
@leonardvandriel Thanks a ton Leonard, you have been really helpful! This is where things seem to be getting even more complex; I see from that example that you use a `TensorCPU` which I can replace with a `TensorCUDA` following all the pre-processing on the GPU.
However, when I tried to add `init_model.mutable_device_option()->set_device_type(CUDA);` in my code, the actual application SEGFAULTs. My guess it is that option that sets the network to use the GPU instead of the CPU?

It is my understanding that not all models are built for CUDA and some will only run on CPU?
If that is the case, then surely there is a way to find if the model(s) support CUDA?
I'm starting to think that in addition to your tutorials, a few wrappers for `caffe2` might be needed.

**Comment by leonardvandriel** [comment]
Hi @alexge233, SEGFAULTs usually indicate a bug or at least the absence of proper input checking. Have you tried to run the application with `lldb` (macOS) or `gdb` (Ubuntu) to see where things go wrong? And which Caffe2 commit are you on? Perhaps it all runs fine on the last v0.8.1 tag.

And what do you exactly mean by "not all models are built for CUDA"? If an operator does not have a CUDA implementation, it should give a proper error message.

**Comment by alexge233** [comment]
@leonardvandriel I didn't put it under gdb, but what surprises me is that removing the line `init_model.mutable_device_option()->set_device_type(CUDA);` stops it from happening.
So if I set the device type to CUDA, and use the code you posted, I should be able to forwrd propagate a CUDA tensor in the network using CUDA only?

I'll have a look next week and see what went wrong.
Also, I was under the impression that some models are trained for CPU and some for CUDA, is this wrong? Not that it is relevant to the question, but useful to know.

**Comment by alexge233** [comment]
The exact error I got is an assertion triggered during my unit test (catch.pp):

```
FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  [enforce fail at predictor.cc:27] blob->template IsType<TensorCPU>(). Blob is
  not a CPU Tensor: data 
```

It seems like I've got to abandon the use of `Predictor` and move on to something else.

**Comment by leonardvandriel** [comment]
The idea is that for  every operator there are multiple implementations, one for each type of hardware (referred to as "device"). So by specifying `set_device_type(CUDA)`, you make the complete model run on the GPU, while if you leave it out it will run on the CPU. Operators usually require input tensors to reside on the hardware they are on, so that's why you get the error `not a CPU Tensor: data`. There are `Copy` operators that copy to and from the GPU, but usually you won't need those. Indeed `Predictor` is not GPU-friendly and it a simple convenience class anyway; just implement that yourself (see `mnist.cc` `predict_example()`).

As for the crash: Adding `set_device_type(CUDA)` runs the complete model on a different hardware using completely different implementation of operators for that hardware. It's not uncommon that models work fine without it, on CPU, and then crash when you add it. It is often an indication that something is wrong in your code, so usually leads to an improvement overall. That being said, I've had some weird crashes with changes in Caffe2 in the past weeks, which they now seem to have solved, so make sure to update to the latest Caffe2 master.

**Comment by alexge233** [comment]
@leonardvandriel Leonard thanks so much, I wouldn't have figured this out without your help.
I've managed to construct, load and execute CUDA versions of both a manually created tensor and the networks. Sadly the `Copy` operators caffe2 has are very unclear as to how they are to be used. For example, trying to copy from cv::cuda::GpuMat has a missing template specialisation. Trying to get help from the caffe2 developers has been very hard.
My last two questions are the following:

1. I see that some models name their output layer differently. I know I can get the blob names, but what is the safest way to get the output blob name? looking into the "predict" file?
2. I see a `RunNet` and `RunNetOnce` and the only documentation on it seems to be for Python (https://caffe2.ai/docs/tutorial-basics-of-caffe2.html). I intend to keep the network alive for the duration of my application (it will be used thousands of times). Should I be using `RunNet` instead? Do you have any examples of how I should be using it?

Again, tons of thanks; I will upload a full example if you want me to, showing how all of the pre-processing can be done using the GPU only.

**Comment by alexge233** [comment]
You know what's driving me crazy?
1. I've looked at the mnist.cc example, and I even went through the predictor.cc code in detail
2. Every time I run the `Predictor` examples using the CPU I get what one would consider "correct" results
3. Every time I run the code which uses CUDA and not `Predictor` I get what appear to be random or "wrong" results.
I've decided to upload the code in a repository, and hopefully make it work as best as possible.
Right now I've committed only the pre-processing methods, but soon I'll add more.
The methodology I'm following is:

```
caffe2::NetDef run_net;
caffe2::NetDef init_net;
caffe2::Workspace workspace;

squeezenet::squeezenet(std::string init_file,
                                        std::string run_file)
: workspace("squeezenet")
{
CAFFE_ENFORCE(ReadProtoFromFile(init_file, &init_net));
CAFFE_ENFORCE(ReadProtoFromFile(pred_file, &run_net));
init_net.mutable_device_option()->set_device_type(caffe2::CUDA);
run_net.mutable_device_option()->set_device_type(caffe2::CUDA);
}

std::vector<result> squeezenet::predict(caffe2::TensorCUDA input)
{
    auto data = workspace.GetBlob("data")->GetMutable<caffe2::TensorCUDA>();
    assert(data);
    data->ResizeLike(input);
    data->ShareData(input);
    workspace.RunNet(run_net.name());
    auto out = workspace.GetBlob("softmaxout");
    assert(out);
    auto softmax = out->Get<caffe2::TensorCUDA>();
    return label(caffe2::TensorCPU(softmax));
}

**Comment by alexge233** [comment]
I've managed to solve my problem and answer my questions.
Again many thanks, I'll post a repository I've built for this, feel free to copy-paste if you need!

**Comment by alexge233** [comment]
@leonardvandriel I've uploaded the GPU stuff using OpenCV here: https://github.com/alexge233/espresso-maker
Feel free to copy-paste :-)

**Comment by leonardvandriel** [comment]
Awesome, thanks for sharing the code. I'll see if I can incorporate in the mnist or imagenet examples.
